python/hail/experimental/ldscore.py

```diff
@@ -71,107 +69,88 @@ def ld_score(entry_expr, annotation_exprs,
-                break
-
-        stops[i] = k
-
-    R2_adj_sparse = R2_adj.sparsify_row_intervals([int(x) for x in starts],
-                                                  [int(x) for x in stops])
-    L2 = R2_adj_sparse @ A
-
-    tmp_bm_path = new_temp_file()
-    tmp_tsv_path = new_temp_file()
-
-    L2.write(tmp_bm_path, force_row_major=True)
-    BlockMatrix.export(tmp_bm_path, tmp_tsv_path)
-
-    ht_scores = hl.import_table(tmp_tsv_path, no_header=True, impute=True)
+    N = mt.count_cols()
+
+    R2 = hl.row_correlation(entry_expr, block_size) ** 2
+    R2_adj = (N - 1.0)/(N - 2.0) * R2 - 1.0/(N - 2.0)
+
+    starts, stops = hl.linalg.utils.locus_windows(locus_expr,
+                                                  radius,
+                                                  coord_expr)
+    R2_adj_sparse = R2_adj.sparsify_row_intervals(starts, stops)
+
+    tmp_R2_adj_sparse_file = new_temp_file()
+    R2_adj_sparse.write(tmp_R2_adj_sparse_file, force_row_major=True)
+    R2_adj_sparse = BlockMatrix.read(tmp_R2_adj_sparse_file)
+
+    if not annotation_exprs:
+        cols = ['univariate']
+        col_idxs = {0: 'univariate'}
+        L2 = R2_adj_sparse.sum(axis=1)
+    else:
+        ht = mt.select_rows(*wrap_to_list(annotation_exprs)).rows()
+        ht = ht.annotate(univariate=hl.literal(1.0))
+        names = [x for x in ht.row if x not in ht.key]
+
+        ht_union = hl.Table.union(
+            *[(ht.annotate(name=hl.str(x),
+                           value=hl.float(ht[x]))
+                 .select('name', 'value')) for x in names])
+        mt_annotations = ht_union.to_matrix_table(
+            row_key=[x for x in ht_union.key],
+            col_key=['name'])
+
+        cols = mt_annotations['name'].collect()
+        col_idxs = {i: cols[i] for i in range(len(cols))}
+
+        A = BlockMatrix.from_entry_expr(mt_annotations.value)
+        tmp_A_file = new_temp_file()
+        BlockMatrix.write_from_entry_expr(mt_annotations.value, tmp_A_file)
+
+        A = BlockMatrix.read(tmp_A_file)
+        L2 = R2_adj_sparse @ A
+
+    tmp_L2_bm_file = new_temp_file()
+    tmp_L2_tsv_file = new_temp_file()
+    L2.write(tmp_L2_bm_file, force_row_major=True)
+    BlockMatrix.export(tmp_L2_bm_file, tmp_L2_tsv_file)
+
+    ht_scores = hl.import_table(tmp_L2_tsv_file, no_header=True, impute=True)
     ht_scores = ht_scores.add_index()
     ht_scores = ht_scores.key_by('idx')
     ht_scores = ht_scores.rename({'f{:}'.format(i): col_idxs[i]
                                   for i in range(len(cols))})
 
-    ht_variants = mt.rows()
-    ht_variants = ht_variants.drop(
-        *[x for x in ht_variants.row if x not in variant_key])
-    ht_variants = ht_variants.add_index()
-    ht_variants = ht_variants.key_by('idx')
-
-    ht_scores = ht_variants.join(ht_scores, how='inner')
-    ht_scores = ht_scores.key_by('locus')
-    ht_scores = ht_scores.drop('alleles', 'idx')
+    ht = mt.select_rows(__locus=locus_expr).rows()
```

Comment:
`locus_expr` may not be on `mt`.  shouldn't you use `locus` on mt? we already assumed alignment right?

Reply:
Where would `locus_expr` be coming from if not `mt`? All of the exprs (`entry_expr`, `locus_expr`, `coord_expr`, `annotation_exprs`) should be coming from the same `mt`. How do I check that?

Reply:
It could be coming from mt.rows(). But yeah, its reasonable for you to require they all come from the same matrix table, and you can check by checking they all have the same expr._indices.source and throwing an error message if not. Say in the docs that you require all the expression parameters to be on the same matrix table.

Reply:
I added a check to ensure `entry_expr`, `locus_expr`, `coord_expr`, and `annotation_exprs` are all coming from the same matrix table.